src/logging/container.ts

```diff
@@ -0,0 +1,98 @@
+import * as es from 'event-stream';
+import { EventEmitter } from 'events';
+import * as _ from 'lodash';
+import * as Stream from 'stream';
+import StrictEventEmitter from 'strict-event-emitter-types';
+
+import Docker from '../lib/docker-utils';
+
+export interface ContainerLog {
+	message: string;
+	timestamp: number;
+	isStdout: boolean;
+}
+
+interface LogsEvents {
+	log: ContainerLog;
+	closed: void;
+	error: Error;
+}
+
+type LogsEventEmitter = StrictEventEmitter<EventEmitter, LogsEvents>;
+
+export class ContainerLogs extends (EventEmitter as {
+	new (): LogsEventEmitter;
+}) {
+	public constructor(public containerId: string, private docker: Docker) {
+		super();
+	}
+
+	public async attach(lastSentTimestamp: number) {
+		const logOpts = {
+			follow: true,
+			timestamps: true,
+			since: Math.floor(lastSentTimestamp / 1000),
+		};
+		const stdoutLogOpts = _.merge({ stdout: true, stderr: false }, logOpts);
+		const stderrLogOpts = _.merge({ stderr: true, stdout: false }, logOpts);
+
+		const container = this.docker.getContainer(this.containerId);
+		const stdoutStream = await container.logs(stdoutLogOpts);
+		const stderrStream = await container.logs(stderrLogOpts);
+
+		[[stdoutStream, true], [stderrStream, false]].forEach(
+			([stream, isStdout]: [Stream.Readable, boolean]) => {
+				stream
+					.on('error', err => {
+						this.emit(
+							'error',
+							new Error(`Error on container logs: ${err} ${err.stack}`),
+						);
+					})
+					.pipe(es.split())
+					.on('data', (logBuf: Buffer | string) => {
+						if (_.isString(logBuf)) {
+							logBuf = Buffer.from(logBuf);
+						}
+						const logMsg = ContainerLogs.extractMessage(logBuf);
+						if (logMsg != null) {
+							this.emit('log', _.merge({ isStdout }, logMsg));
```

Comment:
```typescript
this.emit('log', { isStdout, ...logMsg });
```

Reply:
That's equivalent to `_.assign` not to `_.merge` @zvin, that said I expect @CameronDiver actually wants the assign behaviour and not merge behaviour here